### PR TITLE
Move parameters to environment variables

### PR DESCRIPTION
**Pull request by Walz**
# Description

To simplify the command and to not expose secrets, the parameters of `brimstone` and `hailstone` are moved into environment variables.

The README.md is updated to reflect this change.

### CODE REVIEW

**Comment by pierrelalanne** [comment]
@davidh-cyberark, did you have some time to look at this PR ? 🙂  Thanks a lot!

File: cmd/brimstone/main.go [hunk 1]
```diff
@@ -20,87 +20,79 @@ import (
...
+	TlsSkipVerify bool `env:"TLS_SKIP_VERIFY" envDefault:"false"`
+}
 
-	idtenanturl := flag.String("idtenanturl", "", "PAM config ID tenant URL")
-	pcloudurl := flag.String("pcloudurl", "", "PAM config Privilege Cloud URL")
-	pamuser := flag.String("pamuser", "", "PAM config PAM User")
-	pampass := flag.String("pampass", "", "PAM config PAM Pass")
-	safename := flag.String("safename", "", "PAM config PAM Safe Name")
+func main() {
+	e := echo.New()
```
Comment (davidh-cyberark) [inline cmd/brimstone/main.go:54]: Why was this moved to here?  Prefer to keep this initialization closer to where it is being used.

File: cmd/brimstone/main.go [hunk 2]
```diff
@@ -20,87 +20,79 @@ import (
 	"github.com/labstack/echo/v4/middleware"
 
 	"gorm.io/driver/postgres"
+	"gorm.io/driver/sqlite"
 	"gorm.io/gorm"
+
+	"github.com/caarlos0/env/v6"
```
Comment (davidh-cyberark) [inline cmd/brimstone/main.go]: Is this the latest version?  Looks like v10 might be the latest version.  Recommend using the latest version.

File: cmd/brimstone/main.go [hunk 3]
```diff
@@ -126,23 +118,30 @@ func main() {
 			return len(ggsig) > 0
 		},
 		Validator: func(key string, c echo.Context) (bool, error) {
-			brimstoneapikey, ok := GetAPIKey()
-			return (ok && key == brimstoneapikey), nil
+			return key == cfg.ApiKey, nil
```
Comment (davidh-cyberark) [inline cmd/brimstone/main.go:115]: since the echo context is passed in, and the config was added to the context; recommend to get api key from the context-config, smth like c.Get("config").APIKey

File: cmd/brimstone/main.go [hunk 4]
```diff
@@ -182,11 +172,7 @@ func GGValidator(ggsig string, c echo.Context) (bool, error) {
 	if bbErr != nil {
 		return false, fmt.Errorf("unable to read request body")
 	}
-	webhooktoken, ok := GetGGWebhookToken()
-	if !ok {
-		return false, fmt.Errorf("no webhook token set")
-	}
-	if ok := gg.ValidateGGPayload(ggsig, ggts, webhooktoken, bodyBytes); !ok {
+	if ok := gg.ValidateGGPayload(ggsig, ggts, cfg.GgWebhookToken, bodyBytes); !ok {
```
Comment (davidh-cyberark) [inline cmd/brimstone/main.go:170]: Please add back in the check to ensure that the gg webhook token is set.

File: cmd/brimstone/main.go [hunk 5]
```diff
@@ -159,20 +158,11 @@ func main() {
...
-
-func GetAPIKey() (string, bool) {
-	return os.LookupEnv(API_KEY_VARNAME)
-}
-func GetGGAPIToken() (string, bool) {
-	return os.LookupEnv(GG_API_TOKEN_VARNAME)
-}
-func GetGGWebhookToken() (string, bool) {
-	return os.LookupEnv(GG_WEBHOOK_TOKEN_VARNAME)
+	e.Logger.Fatal(e.Start(net.JoinHostPort("0.0.0.0", strconv.Itoa(int(cfg.Port)))))
```
Comment (davidh-cyberark) [inline cmd/brimstone/main.go]: Can this be simplified by declaring the Port as an "int" in the struct definition?

File: cmd/hailstone/main.go [hunk 6]
```diff
@@ -17,26 +18,34 @@ import (
 	hmsl "github.com/davidh-cyberark/brimstone/pkg/hasmysecretleaked"
 	pam "github.com/davidh-cyberark/brimstone/pkg/privilegeaccessmanager"
 	"github.com/davidh-cyberark/brimstone/pkg/utils"
+
+	"github.com/caarlos0/env/v6"
```
Comment (davidh-cyberark) [inline cmd/hailstone/main.go]: Please check that this is the latest version.  Looks like v10 might be the latest.

File: cmd/hailstone/main.go [hunk 7]
```diff
@@ -17,26 +18,34 @@ import (
...
+	DEBUG   = false
 )
 
-func main() {
-	idtenanturl := flag.String("idtenanturl", "", "PAM config ID tenant URL")
-	pcloudurl := flag.String("pcloudurl", "", "PAM config Privilege Cloud URL")
-	pamuser := flag.String("pamuser", "", "PAM config PAM User")
-	pampass := flag.String("pampass", "", "PAM config PAM Pass")
-	safename := flag.String("safename", "", "PAM config PAM Safe Name")
+type config struct {
```
Comment (davidh-cyberark) [inline cmd/hailstone/main.go:30]: Can the struct from brimstone be re-used?

**Comment by davidh-cyberark** [comment]
@pierrelalanne @Walz I've added my comments. 

Question: I noticed the use of the "env" module, was the "koanf" module considered as well?

File: cmd/brimstone/main.go (hunk 4 above)
Comment (Walz) [reply cmd/brimstone/main.go:170]: It's done by `env` when `config` is initialized because the gg webhook token is required.

File: cmd/brimstone/main.go (hunk 5 above)
Comment (Walz) [reply cmd/brimstone/main.go]: [`JoinHostPort`](https://pkg.go.dev/net#JoinHostPort) expects two `string`, but to validate the port is valid, I've declared `Port` as a `uint16` because a port should be between 0 and 65535. 

I've split this line to make it easier to read.

File: cmd/hailstone/main.go (hunk 7 above)
Comment (Walz) [reply cmd/hailstone/main.go:30]: They do not hold the exactly the same info but share the CyberArk credentials.
I have merged the common part in a struct `BaseConfig`

File: cmd/brimstone/main.go (hunk 1 above)
Comment (Walz) [reply cmd/brimstone/main.go:54]: `e` is used below to log the error if loading the env variables fails.

**Comment by Walz** [comment]
> Question: I noticed the use of the "env" module, was the "koanf" module considered as well?

I did not know/considered "koanf".
"env" give more control and validation for specifically environment variables but "koanf" handles a lot more format.

File: README.md
Comment (davidh-cyberark) [inline README.md:1]: All the bullets were converted from `*` to "-" this is inconsistent with the other documents in the project.  Please update to use `*` for bullets.  the `make docs` rule shows how to use `doctoc` to update the TOC to use `*`

**Comment by davidh-cyberark** [comment]
@Walz
* Please add an example env file.
* Please add an example how to start brimstone using the env file in the README.md

**Comment by Walz** [comment]
I've updated the `README.md` and added a `.env.example` file @davidh-cyberark

**Comment by davidh-cyberark** [comment]
@Walz 

Please move the BaseConfig into pkg/brimstone, it does not belong in utils.   The base config is for the brimstone service, and, since the brimstone pkg is already imported into hailstone, it should be straight forward to update BaseConfig.

**Comment by Walz** [comment]
> @Walz
> 
> Please move the BaseConfig into pkg/brimstone, it does not belong in utils. The base config is for the brimstone service, and, since the brimstone pkg is already imported into hailstone, it should be straight forward to update BaseConfig.

Done!

**Comment by pierrelalanne** [comment]
Hi @davidh-cyberark,
Any chance we can merge this PR ? We're documenting how to deploy Brimstone and would like to have consistent instructions between our content and the repo.  
Thanks
